src/neptune/internal/operation_processors/lazy_operation_processor_wrapper.py

```diff
@@ -0,0 +1,112 @@
+#
+# Copyright (c) 2024, Neptune Labs Sp. z o.o.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import annotations
+
+__all__ = ("LazyOperationProcessorWrapper",)
+
+from pathlib import Path
+from typing import (
+    Any,
+    Callable,
+    Optional,
+    TypeVar,
+)
+
+from neptune.core.components.abstract import Resource
+from neptune.core.components.operation_storage import OperationStorage
+from neptune.internal.operation import Operation
+from neptune.internal.operation_processors.operation_processor import OperationProcessor
+
+RT = TypeVar("RT")
+
+
+def trigger_evaluation(method: Callable[..., RT]) -> Callable[..., RT]:
+    def _wrapper(self: LazyOperationProcessorWrapper, *args: Any, **kwargs: Any) -> RT:
+        if self._operation_processor is None:
+            self._operation_processor = self._operation_processor_getter()
+            if self._post_trigger_side_effect is not None:
+                self._post_trigger_side_effect()
+        return method(self, *args, **kwargs)
+
+    return _wrapper
+
+
+def noop_if_not_triggered(method: Callable[..., RT]) -> Callable[..., RT]:
+    def _wrapper(self: LazyOperationProcessorWrapper, *args: Any, **kwargs: Any) -> RT:
+        if self._operation_processor is not None:
+            return method(self, *args, **kwargs)
+
+    return _wrapper
+
+
+class LazyOperationProcessorWrapper(OperationProcessor):
+    def __init__(
+        self,
+        operation_processor_getter: Callable[[], OperationProcessor],
+        post_trigger_side_effect: Optional[Callable[[], Any]] = None,
+    ):
+        self._operation_processor_getter = operation_processor_getter
+        self._post_trigger_side_effect = post_trigger_side_effect
+        self._operation_processor: OperationProcessor = None  # type: ignore
```

Comment:
Shouldn't it be `Optional[OperationProcessor]`?

Reply:
Or was it to simplify the implementation?

Reply:
It should be `Optional[OperationProcessor]` but then I can not call `self._operation_processor.some_method()` as from typchecker perspective `self._operation_processor` might be `None`. We know that wont be the case becouse of all the decorators wrapping all of methods, but I am not aware of any clean way to implement this kind of type guard, so I just went with  typecheck supress in one line

Reply:
Sure, I agree. Just wanted to know the rationale :wink:. Thanks!